### DISCUSSION TITLE
Creating a virtual Zigbee devices for test

### DISCUSSION THREAD

**Discussion by anuraja-sivasubramanian** [zigpy/zigpy#1459] [General]
Is there a way to create a virtual Zigbee NCP without using a real radio? I couldn’t find any specific examples of creating a virtual Coordinator and end device using a mock radio, possibly via a TCP socket instead of a real device. Any information on configuring a virtual Zigbee device would be greatly appreciated.

**Comment by Hedda** [comment]
For inspiration in regards to creating Zigbee End Device (or Zigbee Router devices), maybe you could look at the new Zigbee Wrapper for Arduino that Espressif's employee/developer P-R-O-C-H-Y has started working on as simple Zigbee API withon Arduino (for arduino-esp32 and depending on Espressif's esp-zigbee-sdk) to allow creation of Zigbee devices using Arduino. Parhaps same API could be used as framework? As I understand it his Zigbee Wrapper for Arduino is something that will make it easier to create your own Zigbee devices (using Espressif's existing ESP3-H2 and ESP3-C6 or the upcoming ESP3-H4 and ESP3-C5 microcontrollers). See:

* https://www.cnx-software.com/2024/08/23/arduino-core-for-esp32-gets-a-zigbee-wrapper-library/

and

* https://github.com/espressif/arduino-esp32/issues/10135

In related news; luar123, tomaszduda23, and Muk911 have all three (independently of each other) begun working on Zigbee device support with YAML config for ESPHome with the the concept that all  Zigbee device variables can be configurated via YAML in ESPHome. Check out:

* https://github.com/luar123/zigbee_esphome
* https://github.com/tomaszduda23/esphome/pull/2
* https://github.com/Muk911/esphome

PS: Also check out this project to control any Arduino program with ESPHome:

* https://github.com/DirkHeinke/arduinoToESPHome-guide
